File: easy/1071_greatest_common_divisor_of_strings/solution.py
import math


class Solution:
    def gcdOfStrings(self, str1: str, str2: str) -> str:
        pivot = math.gcd(len(str1), len(str2))

        if str1[:pivot] != str2[:pivot]:
            return ""

        if (
            str1.count(str1[:pivot]) != len(str1) // pivot
            or str2.count(str2[:pivot]) != len(str2) // pivot
        ):
            return ""

        return str1[:pivot]

File: easy/1071_greatest_common_divisor_of_strings/solution.cpp
#include <numeric>
#include <string>

class Solution {
public:
  std::string gcdOfStrings(std::string str1, std::string str2) {
    const size_t pivot = std::gcd(str1.size(), str2.size());
    const std::string substr = str1.substr(0, pivot);

    for (const std::string &str : {str1, str2}) {
      for (size_t pos = 0; pos < str.size(); pos += pivot) {
        if (str.find(substr, pos) != pos)
          return "";
      }
    }

    return substr;
  }
};
